### PR TITLE
[DEV-2459] add transaction uid

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
- Add migration
- Synchronize after merge

File: src/subdomains/supporting/payment/services/transaction.service.ts [hunk 1]
```diff
@@ -10,37 +15,45 @@ import { TransactionRepository } from '../repositories/transaction.repository';
...
-  async create(dto: CreateTransactionDto): Promise<Transaction | undefined> {
+  async create(
+    dto: CreateTransactionDto,
+    sourceEntity: BankTx | CryptoInput | CheckoutTx | RefReward,
+  ): Promise<Transaction | undefined> {
     const entity = this.repo.create(dto);
 
+    const hash = Util.createHash(
+      `${sourceEntity.id}` + entity.created + entity.sourceType + sourceEntity.created,
+    ).toUpperCase();
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction.service.ts:26]: Does the entity already have a `created` date? I think I would take only fields from source entity.

File: src/subdomains/supporting/payment/services/transaction.service.ts [hunk 2]
```diff
@@ -10,37 +15,45 @@ import { TransactionRepository } from '../repositories/transaction.repository';
...
+    entity.uid = `${hash.slice(0, 8)}-${hash.slice(8, 16)}-${hash.slice(16, 24)}`;
+
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction.service.ts:28]: 24 digits seems to be a bit long for me 😁  and do you see some benefit using these dashes?

File: src/subdomains/core/history/controllers/transaction.controller.ts [hunk 3]
```diff
@@ -80,6 +83,27 @@ export class TransactionController {
     return this.getHistoryData(query, ExportType.COMPACT, res);
   }
 
+  @Get('uid/:uid')
+  @ApiOkResponse({ type: TransactionDto })
+  async getSingleTransaction(@Param('uid') uid: string): Promise<TransactionDto> {
```
Comment (davidleomay) [inline src/subdomains/core/history/controllers/transaction.controller.ts:88]: Should we just use `@Get(':uid')` or do you want to reserve this for a future call with the "real" ID?

File: src/subdomains/core/history/controllers/transaction.controller.ts [hunk 4]
```diff
@@ -80,6 +83,27 @@ export class TransactionController {
...
+    const transaction = await this.transactionService.getTransactionByUid(uid, { buyCrypto: true, buyFiat: true });
+    if (!transaction) throw new NotFoundException('Transaction not found');
+    if (!transaction.txTarget) throw new BadRequestException('Unsupported transaction type');
+
+    switch (transaction.txTarget.constructor) {
```
Comment (davidleomay) [inline src/subdomains/core/history/controllers/transaction.controller.ts:93]: Remove the check and change the switch to
`switch (transaction.txTarget?.constructor) {`

File: src/subdomains/core/history/controllers/transaction.controller.ts [hunk 5]
```diff
@@ -80,6 +83,27 @@ export class TransactionController {
...
+      case BuyCrypto:
+        const buyCryptoExtended = await this.buyCryptoWebhookService.extendBuyCrypto(transaction.buyCrypto);
+        return TransactionDtoMapper.mapBuyCryptoTransaction(buyCryptoExtended);
+
+      case BuyFiat:
+        const buyFiatExtended = await this.buyFiatService.extendBuyFiat(transaction.buyFiat);
+        return TransactionDtoMapper.mapBuyFiatTransaction(buyFiatExtended);
+
+      default:
+        throw new BadRequestException('Unsupported transaction type');
```
Comment (davidleomay) [inline src/subdomains/core/history/controllers/transaction.controller.ts:103]: What about referral transactions?

File: src/subdomains/core/history/controllers/transaction.controller.ts [hunk 6]
```diff
@@ -80,6 +83,27 @@ export class TransactionController {
...
+    }
+  }
```
Comment (davidleomay) [inline src/subdomains/core/history/controllers/transaction.controller.ts:105]: Ah, this code is already there below, maybe you can reuse it?

**Review by davidleomay — APPROVED** [review]

**Pull request merged**